### DISCUSSION TITLE
Stuck on timeout error

### DISCUSSION THREAD

**Discussion by vitarkah** [argoproj/argo-cd#11802] [Q&A]
Trying to add EKS cluster A to ArgoCD running in Cluster B (same account)

Get "https://aaaaa.gr7.us-east-2.eks.amazonaws.com/version?timeout=32s": getting credentials: exec: executable argocd-k8s-auth failed with exit code 20 (Client.Timeout exceeded while awaiting headers)

Looked at Subnet ACL, Security Groups & IP Access (EKS is public & private with allowed IP; so allowing Argo CD / EKS NAT Elastic IP in the cluster A)

I am also able to CURL from Argo CD EC2 Instance (from Nodegroup) to the URL above

Any pointers / idea would be super helpful. At this point I am totally stuck.

**Comment by vitarkah** [comment]
turned out to be a problem with the one of the policies.  Addressing the policy took care care of the issue

**Reply by isihu** [reply]
Could you elaborate on that, please ? I'm having the same issue. Which policy turned out to be faulty?

**Reply by jorgelon** [reply]
yes please. same error here

**Comment by vitarkah** [comment]
The pattern setup closely follows the outline provided here: https://www.modulo2.nl/blog/argocd-on-aws-with-multiple-clusters.  The script that set up the clusters, failed on argocd-deployer role set up. Without that role (or an equivalent in my case) it was just not working.  Once that was set up with right policies, things just worked fine.

**Reply by jorgelon** [reply]
mmm i have the same problem. what policies did you add. thx

**Comment by isihu** [comment]
Funny, I was following the same tutorial. 
For anyone reading this:
My mistake was using the **same** aws role for the argocd service accounts and for the deployment. In the tutorial, ArgoCD assumes the deployment role from its irsa roles. In my case, that means that Argo tries to assume the role **it's already in** leading to this cryptic error. 
Due to my use case I have to use the same aws role. This can be possible by allowing the role be assumed from itself:

```
        {
            "Sid": "",
            "Effect": "Allow",
            "Action": [
                "sts:AssumeRole"
            ],
            "Principal": {
                "AWS": "arn:aws:sts::{ACCOUNT_ID}:role/{ROLE}"
            }
        }
```

**Reply by rezonick** [reply]
Thank you! had the same issue :)

**Reply by moonape1226** [reply]
Same issue here. Thanks for the solution!

**Reply by eagleusb** [reply]
Same here, I did use the [terraform module](https://github.com/terraform-aws-modules/terraform-aws-iam/blob/master/modules/iam-role-for-service-accounts-eks) to create the IRSA and setup the [self assume role policy](https://github.com/terraform-aws-modules/terraform-aws-iam/blob/master/modules/iam-role-for-service-accounts-eks/main.tf#L18C20-L18C42) which worked 👍🏽

**Comment by korbin** [comment]
Both the `argocd-application-controller` and `argocd-server` IAM roles apparently need role delegation. Until adding the former this would consistently timeout for me, regardless of the assume role config.

**Comment by s33dunda** [comment]
Wow thanks all. I was expecting to get an auth error on first attempt but got a timeout instead. What a red herring!

**Comment by AnubhavSabharwa** [comment]
I am able to get the token via

argocd-k8s-auth aws --cluster-name test-dev --role-arn arn:aws:iam
Able to get the token but still on argocd the cluster addition got failed.

Below error:
**/version?timeout=32s": getting credentials: exec: executable argocd-k8s-auth failed with exit code 20 (Client.Timeout exceeded while awaiting headers)**

Please help me out on this

**Reply by magzim21** [reply]
I had this error. Turned out that I did not configure the current principal to assume the role. 
The error message is misleading.

**Reply by rcfja** [reply]
@magzim21 how do I configure the current principal to assume the role? Is that the "kube-system/aws-auth configmap" from here? https://argo-cd.readthedocs.io/en/stable/operator-manual/declarative-setup/

**Reply by magzim21** [reply]
no, the iam principal of argocd's service account (pod identity)

**Comment by AnubhavSabharwa** [comment]
Is this working with POD identity?
I have setup everything however, still facing the timeout error.

**Reply by shanlashari** [reply]
+1

**Reply by Utwo** [reply]
Check my comment from here https://github.com/argoproj/argo-cd/issues/17064#issuecomment-2636509477

**Comment by aagarwalla-fx** [comment]
One thing that I missed was **restarting** **argocd-server** and **application-controller** after adding the EKS role annotations in their service accounts.
